src/providers/postgres/qgspostgresconn.cpp

```diff
@@ -1249,13 +1250,21 @@ QString QgsPostgresConn::postgisVersion() const
 bool QgsPostgresConn::setSessionRole( const QString &sessionRole )
 {
   if ( !sessionRole.isEmpty() )
+  {
+    mSessionRoleChanged = true;
     return LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET ROLE %1" ).arg( quotedValue( sessionRole ) ) );
+  }
```

Comment:
I think instead of the bool mSessionRoleChanged for tracking the session role, we're better off going for a QString `mCurrentSessionRole`. And then this would change to 

```suggestion
  if ( sessionRole != mCurrentSessionRole )
  {
    mCurrentSessionRole = sessionRole;
    return LoggedPQexecNR( "QgsPostgresConn", QStringLiteral( "SET ROLE %1" ).arg( quotedValue( sessionRole ) ) );
  }
```

and resetSessionRole would check if `!mCurrentSessionRole.isEmpty()`

That'd save some unnecessary reset role calls even when the roles are in use.